core/src/main/scala/filodb.core/downsample/ChunkDownsampler.scala

```diff
@@ -0,0 +1,272 @@
+  require(colIds.length == 2, s"AvgSc downsample requires column ids of sum and count. Got ${colIds.length}")
+  override val name: String = "dAvgSc"
+  val sumCol = colIds(0)
+  val countCol = colIds(1)
+  override def downsampleChunk(part: TimeSeriesPartition,
+                               chunkset: ChunkSetInfo,
+                               startRow: Int,
+                               endRow: Int): Double = {
+    val sumVecPtr = chunkset.vectorPtr(sumCol)
+    val sumColReader = part.chunkReader(sumCol, sumVecPtr)
+    val cntVecPtr = chunkset.vectorPtr(countCol)
+    val cntColReader = part.chunkReader(countCol, cntVecPtr)
+    val sumSum = sumColReader.asDoubleReader.sum(sumVecPtr, startRow, endRow)
+    val sumCount = cntColReader.asDoubleReader.sum(cntVecPtr, startRow, endRow)
+    sumSum / sumCount
+  }
+}
+
+/**
+  * Downsamples by calculating average of values from one column
+  */
+case class AvgDownsampler(override val colIds: Array[Int]) extends DoubleChunkDownsampler {
+  require(colIds.length == 1, s"Avg downsample requires one column id with data to average. Got ${colIds.length}")
+  override val name: String = "dAvg"
+  override def downsampleChunk(part: TimeSeriesPartition,
+                               chunkset: ChunkSetInfo,
+                               startRow: Int,
+                               endRow: Int): Double = {
+    val vecPtr = chunkset.vectorPtr(colIds(0))
+    val colReader = part.chunkReader(colIds(0), vecPtr)
+    val sum = colReader.asDoubleReader.sum(vecPtr, startRow, endRow)
+    val count = colReader.asDoubleReader.count(vecPtr, startRow, endRow)
+    sum / count
+  }
+}
+
+/**
+  * Downsamples by selecting the last timestamp in the downsample period.
+  */
+case class TimeDownsampler(override val colIds: Array[Int]) extends TimestampChunkDownsampler {
+  require(colIds.length == 1, s"Time downsample requires only one column. Got ${colIds.length}")
+  override val name: String = "tTime"
+  def downsampleChunk(part: TimeSeriesPartition,
+                      chunkset: ChunkSetInfo,
+                      startRow: Int,
+                      endRow: Int): Long = {
+    val vecPtr = chunkset.vectorPtr(colIds(0))
+    val colReader = part.chunkReader(colIds(0), vecPtr).asLongReader
+    colReader.apply(vecPtr, endRow)
+  }
+}
+
+object ChunkDownsampler {
+
+  /**
+    * Parses single downsampler from string notation such as
+    * "name(4@1)" where "name" is the downsampler name, 4 & 1 are the column IDs to be used by the function
+    */
+  def downsampler(strNotation: String): ChunkDownsampler = {
+    val parts = strNotation.split("[(@)]")
+    // TODO possibly better validation of string notation
+    require(parts.size >= 2, s"Downsampler '$strNotation' does not have downsampler name and column id. ")
+    val name = parts(0)
+    val colIds = parts.drop(1).map(_.toInt)
+    name match {
+      case "dMin"   => MinDownsampler(colIds)
```

Comment:
I suggest you use the `Enum` / `EnumEntry` types/classes (like ColumnType has).  It automates / eliminates the code like you have below, makes it much nicer and less prone to error.